src/libbson/src/bson/bson-dsl.h

```diff
@@ -0,0 +1,1336 @@
+      uint32_t len;                                                     \
+      bson_type_t typ = bson_iter_type (&bsonVisitContext.iter);        \
+      if (typ == BSON_TYPE_ARRAY)                                       \
+         bson_iter_array (&bsonVisitContext.iter, &len, &data);         \
+      else if (typ == BSON_TYPE_DOCUMENT)                               \
+         bson_iter_document (&bsonVisitContext.iter, &len, &data);      \
+      else {                                                            \
+         _bsonDSLDebug ("Ignoring doc() for non-document/array value"); \
+         break;                                                         \
+      }                                                                 \
+      bson_t inner;                                                     \
+      bson_init_static (&inner, data, len);                             \
+      _bsonParse (inner, __VA_ARGS__);                                  \
+   } while (0);
+
+#define _bsonVisitOperation_continue _bvContinue = true
+#define _bsonVisitOperation_break _bvBreak = _bvContinue = true
+#define _bsonVisitOperation_require(Predicate)                    \
+   _bsonDSL_begin ("require(%s)", _bsonDSL_str (Predicate));      \
+   if (!bsonPredicate (Predicate)) {                              \
+      bsonParseError =                                            \
+         "Element requirement failed: " _bsonDSL_str (Predicate); \
+   }                                                              \
+   _bsonDSL_end
+
+#define _bsonVisitOperation_error(S) bsonParseError = (S)
+
+#define _bsonVisit_applyOp(P, _const, _count)            \
+   do {                                                  \
+      if (!_bvContinue && !_bvHalt && !bsonParseError) { \
+         _bsonVisitOperation_##P;                        \
+      }                                                  \
+   } while (0);
+
+#define _bsonParse(Doc, ...)                                                   \
+   do {                                                                        \
+      /* Reset the context */                                                  \
+      struct _bsonVisitContext_t _bpCtx = {                                    \
+         .doc = &(Doc),                                                        \
+         .parent = &bsonVisitContext,                                          \
+      };                                                                       \
+      _bsonVisitContextThreadLocalPtr = &_bpCtx;                               \
+      /* Keep track of which elements have been visited based on their index*/ \
+      uint64_t _bpVisitBits_static[4] = {0};                                   \
+      BSON_MAYBE_UNUSED uint64_t *_bpVisitBits = _bpVisitBits_static;          \
+      BSON_MAYBE_UNUSED int _bpNumVisitBitInts =                               \
+         sizeof _bpVisitBits_static / sizeof (uint64_t);                       \
```

Comment:
```suggestion
      BSON_MAYBE_UNUSED size_t _bpNumVisitBitInts =                            \
         sizeof (_bpVisitBits_static) / sizeof (*_bpVisitBits_static);         \
```
`_bpNumVisitBitInts` should have type `size_t`.